src/coreclr/jit/inductionvariableopts.cpp

```diff
@@ -1147,6 +1193,707 @@ bool Compiler::optMakeExitTestDownwardsCounted(ScalarEvolutionContext& scevConte
+    {
+    }
+};
+
+class StrengthReductionContext
+{
+    Compiler*               m_comp;
+    ScalarEvolutionContext& m_scevContext;
+    FlowGraphNaturalLoop*   m_loop;
+    LoopLocalOccurrences&   m_loopLocals;
+
+    ArrayStack<Scev*>         m_backEdgeBounds;
+    SimplificationAssumptions m_simplAssumptions;
+    ArrayStack<CursorInfo>    m_cursors1;
+    ArrayStack<CursorInfo>    m_cursors2;
+
+    void        InitializeSimplificationAssumptions();
+    bool        InitializeCursors(GenTreeLclVarCommon* primaryIVLcl, ScevAddRec* primaryIV);
+    void        AdvanceCursors(ArrayStack<CursorInfo>* cursors, ArrayStack<CursorInfo>* nextCursors);
+    bool        CheckAdvancedCursors(ArrayStack<CursorInfo>* cursors, int derivedLevel, ScevAddRec** nextIV);
+    bool        TryReplaceUsesWithNewPrimaryIV(ArrayStack<CursorInfo>* cursors, ScevAddRec* iv);
+    BasicBlock* FindUpdateInsertionPoint(ArrayStack<CursorInfo>* cursors);
+
+    bool StressProfitability()
+    {
+        return m_comp->compStressCompile(Compiler::STRESS_STRENGTH_REDUCTION_PROFITABILITY, 50);
+    }
+
+public:
+    StrengthReductionContext(Compiler*               comp,
+                             ScalarEvolutionContext& scevContext,
+                             FlowGraphNaturalLoop*   loop,
+                             LoopLocalOccurrences&   loopLocals)
+        : m_comp(comp)
+        , m_scevContext(scevContext)
+        , m_loop(loop)
+        , m_loopLocals(loopLocals)
+        , m_backEdgeBounds(comp->getAllocator(CMK_LoopIVOpts))
+        , m_cursors1(comp->getAllocator(CMK_LoopIVOpts))
+        , m_cursors2(comp->getAllocator(CMK_LoopIVOpts))
+    {
+    }
+
+    bool TryStrengthReduce();
+};
+
+//------------------------------------------------------------------------
+// TryStrengthReduce: Check for legal and profitable derived IVs to introduce
+// new primary IVs for.
+//
+// Returns:
+//   True if any new primary IV was introduced; otherwise false.
+//
+bool StrengthReductionContext::TryStrengthReduce()
+{
+    JITDUMP("Considering " FMT_LP " for strength reduction...\n", m_loop->GetIndex());
+
+    if ((JitConfig.JitEnableStrengthReduction() == 0) &&
+        !m_comp->compStressCompile(Compiler::STRESS_STRENGTH_REDUCTION, 50))
+    {
+        JITDUMP("  Disabled: no stress mode\n");
+        return false;
+    }
+
+    // Compute information about the loop used to simplify SCEVs.
+    InitializeSimplificationAssumptions();
+
+    JITDUMP("  Considering primary IVs\n");
+
+    // We strength reduce only candidates where we see that we'll be able to
+    // remove all uses of a primary IV by introducing a different primary IV.
+    //
+    // The algorithm here works in the following way: we process each primary
+    // IV in turn. For every primary IV, we create a 'cursor' pointing to every
+    // use of that primary IV. We then continuously advance each cursor to the
+    // parent node as long as all cursors represent the same derived IV. Once we
```

Comment:
I'm curious about the "same" derived IV requirement ... if we are say converting from a char array to an int array
```C#
int[] a = ...
char[] b = ...
for (int i = 0; ...)
{
     a[i] = (int) b[i];
}
```
if I understand correctly the current code would not strength reduce because we'd have <L, 0, 4> vs <L, 0, 2>? Seems like we still might want to strength reduce as long as the set of cursors doesn't expand into too many different new IVs? Or say if it is relatively cheap to convert from one family to another (say multiply by 2 which can be done with scaled address mode).

Also wonder if this profitability equation changes if we have post-increment modes (I forget where we ended up on this).

Reply:
> if I understand correctly the current code would not strength reduce because we'd have <L, 0, 4> vs <L, 0, 2>? Seems like we still might want to strength reduce as long as the set of cursors doesn't expand into too many different new IVs? Or say if it is relatively cheap to convert from one family to another (say multiply by 2 which can be done with scaled address mode).

Yeah, that's right. I think you're right we can generalize this: we can keep going if the addrecs have the same 'start' and all step values are divisible by the minimum step value. I'll consider that as a follow-up.

> Also wonder if this profitability equation changes if we have post-increment modes (I forget where we ended up on this).

Currently there's really no heuristics here; we kind of just do strength reduction whenever we have a candidate. I definitely need to sit down and consider more carefully how the transformation affects performance and size (usually it seems to regress size because it requires computing some 'start' value in the preheader).